File: алгоритмы/Brute_forse(Метод_Грубой_силы)/Sort_Сортировка/Insertion_sort.rb
arr = [12, 1, 8, -3, 0, 14, 3, -2, 1]

# ! сортировка вставками
def my_method(arr)
  count = 1

  loop do
    return arr if count >= arr.size

    count2 = count

    loop do
      if arr[count2 - 1] > arr[count2]
        arr[count2 - 1], arr[count2] = arr[count2], arr[count2 - 1]
        count2 -= 1
      else
        break
      end

      break if count2 == 0
    end

    count += 1
  end
end

p my_method(arr)

# ============================================================================

arr2 = [12, 1, 8, -3, 0, 14, 3, -2, 1]

# ! сортировка вставками
def my_method2(arr)
  arr.each_with_index do |value, index|
    if index > 0 && arr[index - 1] > arr[index]
      count = index

      while count > 0
        arr[count - 1], arr[count] = arr[count], arr[count - 1] if arr[count - 1] > arr[count]
          
        count -= 1
      end
    end
  end
end

p my_method2(arr2)

File: алгоритмы/Brute_forse(Метод_Грубой_силы)/Sort_Сортировка/Insertion_sort.js
// ! Сортировка вставками

let arr = [4, 4, 3, 1, 5, 6, 3, 15, 8, 0, -1, -3, -2];

function f1(arr) {
    for (let i = 1; i < arr.length; i++) {
        for (let a = i; arr[a - 1] > arr[a]; a--) {
            [arr[a - 1], arr[a]] = [arr[a], arr[a - 1]];
        }
    }

    return arr;
}

console.log(f1(arr));

// function insertSort(a) {
//     for (var sortedN = 1; sortedN < a.length; sortedN++) {
//         for (var i = sortedN; i > 0 && a[i] < a[i - 1]; i--) {
//             [a[i - 1], a[i]] = [a[i], a[i - 1]];
//         }
//     }

//     return a;
// }

// console.log(insertSort(arr));
